documentation/dataplane/dash-flow-api.md

```diff
@@ -0,0 +1,534 @@
+| SAI_FLOW_ENTRY_BULK_GET_SESSION_ATTR_FIFTH_FLOW_ENTRY_BULK_GET_SESSION_FILTER_ID | @type: `sai_object_id_t` @objects `SAI_OBJECT_TYPE_FLOW_ENTRY_BULK_GET_SESSION_FILTER` | Action set_flow_entry_bulk_get_session_attr parameter THIRD_FLOW_ENTRY_BULK_GET_SESSION_FILTER_ID |
+
+### Protobuf-based flow programming
+
+In addition to the flow state attributes, the flow state can be represented using protobuf for high-efficiency. The attribute of the flow entry is SAI_FLOW_ENTRY_ATTR_FLOW_DATA_PB.
+
+Although the content of both attributes and protobuf may be identical, their applications differ. Attributes enable incremental updates to individual properties, whereas protobuf necessitates a complete update.
+
+```protobuf
+syntax = "proto3";
+
+message SaiDashFlowMetadata {
+  uint32 version = 1; // SAI_FLOW_ENTRY_ATTR_VERSION
+  uint32 dash_flow_action = 2; // SAI_FLOW_ENTRY_ATTR_DASH_FLOW_ACTION
+  uint32 meter_class = 3; // SAI_FLOW_ENTRY_ATTR_METER_CLASS
+  bool is_bidirectional_flow = 4; // SAI_FLOW_ENTRY_ATTR_IS_BIDIRECTIONAL_FLOW
+  uint8 reverse_ip_protocol = 5; // SAI_FLOW_ENTRY_ATTR_REVERSE_IP_PROTOCOL
+  string reverse_src_ip_addr = 6; // SAI_FLOW_ENTRY_ATTR_REVERSE_IP_ADDR for source
+  string reverse_dst_ip_addr = 7; // SAI_FLOW_ENTRY_ATTR_REVERSE_IP_ADDR for destination
+  uint32 reverse_src_l4_port = 8; // SAI_FLOW_ENTRY_ATTR_REVERSE_SRC_L4_PORT
+  uint32 reverse_dst_l4_port = 9; // SAI_FLOW_ENTRY_ATTR_REVERSE_DST_L4_PORT
+  uint32 dest_vnet_vni = 10; // SAI_FLOW_ENTRY_ATTR_DEST_VNET_VNI
+  string underlay_sip = 11; // SAI_FLOW_ENTRY_ATTR_UNDERLAY_SIP
+  string underlay_dip = 12; // SAI_FLOW_ENTRY_ATTR_UNDERLAY_DIP
+  string underlay_smac = 13; // SAI_FLOW_ENTRY_ATTR_UNDERLAY_SMAC
+  string underlay_dmac = 14; // SAI_FLOW_ENTRY_ATTR_UNDERLAY_DMAC
+  bool is_ipv6 = 15; // SAI_FLOW_ENTRY_ATTR_IS_IPV6
+  string dst_mac = 16; // SAI_FLOW_ENTRY_ATTR_DEST_MAC
+  string sip = 17; // SAI_FLOW_ENTRY_ATTR_SIP
+  string dip = 18; // SAI_FLOW_ENTRY_ATTR_DIP
+  string sip_mask = 19; // SAI_FLOW_ENTRY_ATTR_SIP_MASK
+  string dip_mask = 20; // SAI_FLOW_ENTRY_ATTR_DIP_MASK
+}
+```
+
+### Capability
+
+| Attribute Name                                     | Type           | Description                                                  |
+| -------------------------------------------------- | -------------- | ------------------------------------------------------------ |
+| SAI_SWITCH_ATTR_DASH_CAPS_MAX_FLOW_TABLE_COUNT     | `sai_uint32_t` | The max number of flow tables that can be created            |
+| SAI_SWITCH_ATTR_DASH_CAPS_MAX_FLOW_ENTRY_COUNT     | `sai_uint32_t` | The max number of flow entries that can be created in a flow table |
+| SAI_SWITCH_ATTR_DASH_CAPS_BULK_GET_SESSION         | `bool`         | Indicates if it supports bulk get sessions                   |
+| SAI_SWITCH_ATTR_DASH_CAPS_BIDIRECTIONAL_FLOW_ENTRY | `bool`         | Indicates if it supports bi-directional flow entry           |
```

Comment:
we will also need capability for:
- Flow read operation: e.g., query
- Flow write operation: e.g., create/update/delete